packages/format-library/src/link/modal-screens/link-settings-screen.native.js

```diff
@@ -80,12 +83,17 @@ const LinkSettingsScreen = ( {
 		linkValues.isRemovingLink,
 	] );
 
-	const clearFormat = () => {
+	const clearFormat = ( { skipStateUpdates = false } = {} ) => {
 		onChange( { ...value, activeFormats: [] } );
-		setLinkValues( { isActiveLink: false, isRemovingLink: true } );
+		if ( ! skipStateUpdates ) {
+			setLinkValues( { isActiveLink: false, isRemovingLink: true } );
+		}
```

Comment:
Thanks for the note about `skipStateUpdates` David!

I wonder, since we have to pass through lots of functions/components to send skipStateUpdates to `clearFormat()` and `submitLink()`, wdyt about using a different technique to minimize the parameter passing altogether? I'm thinking, would detecting the "component is unmounted" fact and use it in place of `if ( ! skipStateUpdates )` help?

Something like:
```
    const isMounted = useRef( false );
    useEffect( () => {
        isMounted.current = true;

        return () => {
            isMounted.current = false;
        };
    }, [] );
    ...
    ...
    if ( isMounted.current ) {
	setLinkValues( { isActiveLink: false, isRemovingLink: true } );
    }
```

BTW, I only successfully tested the above solution on iOS where I could replicate [the original problem](https://github.com/WordPress/gutenberg/commit/5cccd1e6fad6fbce0f6fddd8320a7a1671b2cb07) (after reverting the 5cccd1e6fad6fbce0f6fddd8320a7a1671b2cb07 to test it) as I wasn't able to replicate on Android either way.

Also, I think that the alternative solution above is probably not the ultimately better solution to the original problem since, I feel that we perhaps are not using the component/modal's lifecycle methods properly... we shouldn't need to worry about state updates during unmounting 🤔

Reply:
> I wonder, since we have to pass through lots of functions/components to send skipStateUpdates to `clearFormat()` and `submitLink()`, wdyt about using a different technique to minimize the parameter passing altogether? I'm thinking, would detecting the "component is unmounted" fact and use it in place of `if ( ! skipStateUpdates )` help?

@hypest yes, the existing code structure necessitates passing parameters between a fair amount of functions. I agree it is not the most concise or easy to follow. 

I am open to refactoring to your proposal. That said, I have two concerns:

1. The proposal somewhat separates the solution from the original problem, and follows [the "anti-pattern" of an `isMounted` flag](https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html). While verbose, the current approach directly addresses the problem by avoiding state modifications within the callback intended to run during component removal. 
2. I wonder if tracking `isMounted` in a separate effect might open the possibility of a race condition between it and the `onHandleClosingBottomSheet` callback invoked when closing the component. 

> BTW, I only successfully tested the above solution on iOS where I could replicate [the original problem](https://github.com/WordPress/gutenberg/commit/5cccd1e6fad6fbce0f6fddd8320a7a1671b2cb07) (after reverting the [5cccd1e](https://github.com/WordPress/gutenberg/commit/5cccd1e6fad6fbce0f6fddd8320a7a1671b2cb07) to test it) as I wasn't able to replicate on Android either way.

It sounds like your testing may nullify my second concern, so that is good news. 👍🏻 

> Also, I think that the alternative solution above is probably not the ultimately better solution to the original problem since, I feel that we perhaps are not using the component/modal's lifecycle methods properly... we shouldn't need to worry about state updates during unmounting 🤔

I am not sure I agree with this. Will you please elaborate on what you mean by that? From my perspective, based on my understanding of React's architecture, we _are_ responsible for avoiding state updates to unmounted components.

Reply:
> I am not sure I agree with this. Will you please elaborate on what you mean by that? From my perspective, based on my understanding of React's architecture, we are responsible for avoiding state updates to unmounted components.

Oh yeah, totally, _we_ are responsible for wrangling the state updates.

What I mean is that we should structure the code in such a way that we nicely capture the link-setting screen lifecycle (maybe by introducing some custom hooks or additional context?) so we don't need to "manually" pass status flags to differentiate code paths.

Perhaps the fact that `submit()` updates some state is weird in the first place, since it seems like it introduces a loop, and we have to break that loop when the link-settings screen closes. After looking at the code flows some more, my impression is that the `linkValues` state is not really state but rather, a way to circumnavigate the hooks and trigger a function call (`submitLink()` or `removeLink()`), and perhaps that's where the code structure issue lies. In any case, this cannot be an issue to deal with in an RN upgrade PR but in a follow up.

All that said, and since I do agree that my `isMounted` proposal above is not good avenue, I wonder if there's an even "simpler" way out in our case: wdyt about using the already available `shouldCloseBottomSheet` flag, instead of introducing the `skipStateUpdates`? My understanding is that that flag has very similar semantics and indeed we just want to avoid setting the state when the modal is closing. A solution based on `shouldCloseBottomSheet` could look like the following, where we actually just need to augment the signature of `clearFormat()`, close to your original solution anyway, just without introducing a new flag:

```diff
diff --git a/packages/format-library/src/link/modal-screens/link-settings-screen.native.js b/packages/format-
index 3cbdd7344b..5917217ca7 100644
--- a/packages/format-library/src/link/modal-screens/link-settings-screen.native.js
+++ b/packages/format-library/src/link/modal-screens/link-settings-screen.native.js
@@ -83,9 +83,11 @@ const LinkSettingsScreen = ( {
                linkValues.isRemovingLink,
        ] );

-       const clearFormat = () => {
+       const clearFormat = ( shouldCloseBottomSheet ) => {
                onChange( { ...value, activeFormats: [] } );
-               setLinkValues( { isActiveLink: false, isRemovingLink: true } );
+               if ( ! shouldCloseBottomSheet ) {
+                       setLinkValues( { isActiveLink: false, isRemovingLink: true } );
+               }
        };

        const submitLink = ( shouldCloseBottomSheet = true ) => {
@@ -130,7 +132,9 @@ const LinkSettingsScreen = ( {
                }
                newAttributes.activeFormats = [];
                onChange( { ...newAttributes, needsSelectionUpdate: true } );
-               setLinkValues( { isActiveLink: true, isRemovingLink: false } );
+               if ( ! shouldCloseBottomSheet ) {
+                       setLinkValues( { isActiveLink: true, isRemovingLink: false } );
+               }

                if ( ! isValidHref( url ) ) {
                        speak(
@@ -151,7 +155,7 @@ const LinkSettingsScreen = ( {
        };

        const removeLink = ( shouldCloseBottomSheet = true ) => {
-               clearFormat();
+               clearFormat( shouldCloseBottomSheet );
                onRemove();
                if ( shouldCloseBottomSheet ) {
                        onClose();
```

Reply:
Commenting to capture a discussion @hypest and I had on a video call regarding the previous comment and this code. 

* We could likely leverage the existing `shouldCloseBottomSheet` rather than introducing an additional `skipStateUpdates` flag. However, the continued (and the original) refactoring of this code feels a bit out of scope of this PR and possibly better performed in a separate PR. 
* The `LinkSettingsScreen` component may be convoluted and a worthy test case for refactoring, identifying common abstractions — possibly custom React Hooks — for this component and others. Is there opportunity to better compose React Hooks in this component to avoid leaking its concerns to a parent? 
* It is likely worthwhile reviewing the relationship this component has with the `onHandleClosingBottomSheet` callback and its own `linkValues` state. Could the former be managed locally during component unmount? Could the latter be refactored away by invoking `submitLink` and `removeLink` directly within event handlers?